digicampipe/utils/disp.py

```diff
@@ -0,0 +1,290 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib as mpl
+from tqdm import trange
+
+
+def disp_eval(A, width, length, cog_x, cog_y,
+              x_offset, y_offset, psi, skewness,
+              size, leakage2, method):
+
+    parvals = A.valuesdict()
+
+    # (Lessard et al., 2001)
+    if method == 1:
+        disp_comp = parvals['A0']*(1 - width/length)
+
+    # eq 2.10 in Lopez Coto, VHE gamma-ray observations of pulsar
+    # wind nebuale ..., but better reference is: (Domingo-Santamaria+, 2005)
+    elif method == 2:
+        A = (parvals['A0'] + parvals['A1'] * np.log10(size)
+             + parvals['A2'] * np.log10(size)**2)
+        B = (parvals['A3'] + parvals['A4'] * np.log10(size)
+             + parvals['A5'] * np.log10(size)**2)
+        eta = (parvals['A6'] + parvals['A7'] * np.log10(size)
+               + parvals['A8'] * np.log10(size)**2)
+        disp_comp = A + B * (width / (length + eta * leakage2))
+
+    # just some test
+    elif method == 3:
+        disp_comp = parvals['A0'] + parvals['A1'] * length/width
+
+    # Kranich and Stark, ICRC 2003
+    elif method == 4:
+        disp_comp = (parvals['A0']
+                     * (1 - width / (length * (1 + parvals['A1'] * leakage2))))
+    # (Luke Riley St Marie 2014) <-- simplified Domingo-Santamaria
+    elif method == 5:
+        fraction = width / (length + parvals['A2']
+                            * leakage2 * np.log10(size))
+        disp_comp = (np.log10(size)
+                     * (parvals['A0'] + parvals['A1'] * (1 - fraction)))
+
+    x_source_comp0 = cog_x + disp_comp*np.cos(psi)  # Two possible solutions
+    y_source_comp0 = cog_y + disp_comp*np.sin(psi)  #
+    x_source_comp1 = cog_x - disp_comp*np.cos(psi)  #
+    y_source_comp1 = cog_y - disp_comp*np.sin(psi)  #
+
+    # Selection of one specific solution according to skewness
+    # - head of the shower (close to the source) gives more signal
+    # - if skewness > 0, head is closer to the center of FOV than tail
+    x_source_comp = []
+    y_source_comp = []
+    for i in range(len(disp_comp)):
+        if skewness[i] > 0:
+            x_source_comp.append(x_source_comp1[i])
+            y_source_comp.append(y_source_comp1[i])
+        else:
+            x_source_comp.append(x_source_comp0[i])
+            y_source_comp.append(y_source_comp0[i])
+
+    residuals = np.array([x_offset-x_source_comp, y_offset-y_source_comp])
+
+    # Multi-dimensional minimization needs residuals as
+    # a simple 1D vector, thus the .flatten() is applied
+    return disp_comp, x_source_comp, y_source_comp, residuals.flatten()
+
+
+def leak_pixels(pix_x, pix_y, image):
+
+    x_rows = np.unique(pix_x)
+    pix_bound = []
+    for x in x_rows:
+        if x < -412 or x > 398:
```

Comment:
what are these limitations for ? if they specific to the camera geometry maybe you can use the module : `utils.Digicam`
Which contains information on pixel positions and neighbors. Although I am not totally sure of what is done here.

Reply:
This part of the code select the two image pixel rings. So, first it takes X coordinates of the columns and keep two on the left and two on the right. The numbers are X coordinates in mm beyond which the 2+2 columns lie. For upper and bottom side of the camera plane just min and max Y coordinates can be used due to pixel geometry.

Reply:
This function first tries to find which pixels are at the border of the camera by computing a mask right ?
then computes the ration between "border" image and "full" image to get the ratio of the two --> leakage ?

If so I think one can include the first part into the `utils.Digicam` class. As I think it can be useful in the future. At init of the object `utils.Digicam` ones should fill an attribute called `camera_borders_mask`

Reply:
Also if you access the camera object you should be able to get from its attribute `geometry` the neighbor pixels matrix. i.e. a matrix with zeros and ones telling for each pixel which pixel is a neighboring pixel. From that you can count how much neighboring pixels each pixel has. If this count is bellow 6 then it should be at border.

```python
cam = utils.Digicam()
geom = cam.geometry
neighbor_matrix = geom.neighbors
n_neighbors = np.sum(neighbor_maxtrix, axis=0)
camera_border_mask = n_neighbors < 6
```

Reply:
Yes, it is for finding two rows of outermost pixel rings. Thanks for the tip, with the use of camera geometry, it is way more elegant. So, I have just pushed reimplementation with the use of geometry, but including the code into Digicam class is maybe too much for me, because handling with these complicated classes, distributed somehow among multiple files, is a bit beyond my python abilities so far :-)